src/ParallelAlgorithms/NonlinearSolver/NewtonRaphson/ElementActions.hpp

```diff
@@ -0,0 +1,508 @@
+  }
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&, bool, size_t> apply(
+      db::DataBox<DbTagsList>& box, tuples::TaggedTuple<InboxTags...>& inboxes,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& array_index, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    // Retrieve reduction data from inbox
+    auto globalization_result = std::move(
+        tuples::get<Tags::GlobalizationResult<OptionsGroup>>(inboxes)
+            .extract(db::get<Convergence::Tags::IterationId<OptionsGroup>>(box))
+            .mapped());
+
+    if (std::holds_alternative<double>(globalization_result)) {
+      if (UNLIKELY(get<logging::Tags::Verbosity<OptionsGroup>>(box) >=
+                   ::Verbosity::Debug)) {
+        Parallel::printf(
+            "%s " + Options::name<OptionsGroup>() + "(%zu): Globalize(%zu)\n",
+            array_index,
+            db::get<Convergence::Tags::IterationId<OptionsGroup>>(box),
+            db::get<NonlinearSolver::Tags::Globalization<
+                Convergence::Tags::IterationId<OptionsGroup>>>(box));
+      }
+
+      // Update the step length
+      db::mutate<NonlinearSolver::Tags::StepLength<OptionsGroup>,
+                 NonlinearSolver::Tags::Globalization<
+                     Convergence::Tags::IterationId<OptionsGroup>>>(
+          make_not_null(&box),
+          [&globalization_result](const gsl::not_null<double*> step_length,
+                                  const gsl::not_null<size_t*>
+                                      globalization_iteration_id) noexcept {
+            *step_length = get<double>(globalization_result);
+            ++(*globalization_iteration_id);
+          });
+      // Continue globalization by taking a step with the updated step length
+      // and checking the residual again
+      constexpr size_t perform_step_index =
+          tmpl::index_of<ActionList,
+                         PerformStep<FieldsTag, OptionsGroup, Label>>::value;
+      return {std::move(box), false, perform_step_index};
+    }
+
+    // At this point globalization is complete, so we proceed with the algorithm
+    auto& has_converged = get<Convergence::HasConverged>(globalization_result);
+
+    db::mutate<Convergence::Tags::HasConverged<OptionsGroup>>(
+        make_not_null(&box),
+        [&has_converged](const gsl::not_null<Convergence::HasConverged*>
+                             local_has_converged) noexcept {
+          *local_has_converged = std::move(has_converged);
+        });
+
+    constexpr size_t this_action_index =
+        tmpl::index_of<ActionList, Globalize>::value;
+    return {std::move(box), false, this_action_index + 1};
+  }
+};
+
+// Jump back to `PrepareStep` to continue iterating or move on if the algorithm
```

Comment:
It's not clear to me if `move on` here means "We are done with Newton-Raphson, it has converged." or some step in the NR algorithm